File: padlock-pin/src/main/java/com/pyamsoft/padlock/pin/PinEntryPresenter.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.pin;

import android.support.annotation.NonNull;
import com.pyamsoft.pydroid.helper.DisposableHelper;
import com.pyamsoft.pydroid.presenter.SchedulerPresenter;
import io.reactivex.Scheduler;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import javax.inject.Inject;
import javax.inject.Named;
import timber.log.Timber;

class PinEntryPresenter extends SchedulerPresenter {

  @NonNull private final PinEntryInteractor interactor;
  @NonNull private Disposable pinEntryDisposable = Disposables.empty();
  @NonNull private Disposable pinCheckDisposable = Disposables.empty();

  @Inject PinEntryPresenter(@NonNull PinEntryInteractor interactor,
      @Named("obs") Scheduler obsScheduler, @Named("sub") Scheduler subScheduler) {
    super(obsScheduler, subScheduler);
    this.interactor = interactor;
  }

  @Override protected void onStop() {
    super.onStop();
    pinCheckDisposable = DisposableHelper.dispose(pinCheckDisposable);
    pinEntryDisposable = DisposableHelper.dispose(pinEntryDisposable);
  }

  public void submit(@NonNull String currentAttempt, @NonNull String reEntryAttempt,
      @NonNull String hint, @NonNull SubmitCallback callback) {
    Timber.d("Attempt PIN submission");
    pinEntryDisposable = DisposableHelper.dispose(pinEntryDisposable);
    pinEntryDisposable = interactor.submitPin(currentAttempt, reEntryAttempt, hint)
        .subscribeOn(getSubscribeScheduler())
        .observeOn(getObserveScheduler())
        .subscribe(event -> {
          boolean creating = (event.type() == PinEntryEvent.Type.TYPE_CREATE);
          if (event.complete()) {
            callback.onSubmitSuccess(creating);
          } else {
            callback.onSubmitFailure(creating);
          }
        }, throwable -> {
          Timber.e(throwable, "attemptPinSubmission onError");
          callback.onSubmitError();
        });
  }

  public void checkMasterPinPresent(@NonNull MasterPinStatusCallback callback) {
    Timber.d("Check if we have a master");
    pinCheckDisposable = DisposableHelper.dispose(pinCheckDisposable);
    pinCheckDisposable = interactor.hasMasterPin()
        .subscribeOn(getSubscribeScheduler())
        .observeOn(getObserveScheduler())
        .subscribe(hasMaster -> {
          if (hasMaster) {
            callback.onMasterPinPresent();
          } else {
            callback.onMasterPinMissing();
          }
        }, throwable -> Timber.e(throwable, "onError checkMasterPinPresent"));
  }

  interface MasterPinStatusCallback {

    void onMasterPinMissing();

    void onMasterPinPresent();
  }

  interface SubmitCallback {

    void onSubmitSuccess(boolean creating);

    void onSubmitFailure(boolean creating);

    void onSubmitError();
  }
}

File: padlock-pin/src/main/java/com/pyamsoft/padlock/pin/PinEntryPresenter.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.pin

import com.pyamsoft.padlock.api.PinEntryInteractor
import com.pyamsoft.padlock.model.pin.ClearPinEvent
import com.pyamsoft.padlock.model.pin.CreatePinEvent
import com.pyamsoft.padlock.model.pin.PinEntryEvent.Clear
import com.pyamsoft.padlock.model.pin.PinEntryEvent.Create
import com.pyamsoft.pydroid.core.bus.EventBus
import com.pyamsoft.pydroid.core.presenter.Presenter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

class PinEntryPresenter @Inject internal constructor(
  private val interactor: PinEntryInteractor,
  private val createPinBus: EventBus<CreatePinEvent>,
  private val clearPinBus: EventBus<ClearPinEvent>
) : Presenter<PinEntryPresenter.View>() {

  override fun onCreate() {
    super.onCreate()
    checkMasterPinPresent()
  }

  private fun checkMasterPinPresent() {
    Timber.d("Check master pin present")
    dispose {
      interactor.hasMasterPin()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe({
            if (it) {
              view?.onMasterPinPresent()
            } else {
              view?.onMasterPinMissing()
            }
          }, { Timber.e(it, "onError checkMasterPinPresent") })
    }
  }

  private fun publish(event: CreatePinEvent) {
    createPinBus.publish(event)
  }

  private fun publish(event: ClearPinEvent) {
    clearPinBus.publish(event)
  }

  fun submit(
    currentAttempt: String,
    reEntryAttempt: String,
    hint: String
  ) {
    dispose {
      interactor.submitPin(currentAttempt, reEntryAttempt, hint)
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .doAfterTerminate { view?.onPinSubmitComplete() }
          .subscribe({
            when (it) {
              is Create -> {
                if (it.complete) {
                  publish(CreatePinEvent(true))
                  view?.onPinSubmitCreateSuccess()
                } else {
                  publish(CreatePinEvent(false))
                  view?.onPinSubmitCreateFailure()
                }
              }
              is Clear -> {
                if (it.complete) {
                  publish(ClearPinEvent(true))
                  view?.onPinSubmitClearSuccess()
                } else {
                  publish(ClearPinEvent(false))
                  view?.onPinSubmitClearFailure()
                }
              }
            }
          }, {
            Timber.e(it, "attemptPinSubmission onError")
            view?.onPinSubmitError(it)
          })
    }
  }

  interface View : MasterPinCallback, SubmitCallback

  interface SubmitCallback {
    fun onPinSubmitCreateSuccess()
    fun onPinSubmitCreateFailure()
    fun onPinSubmitClearSuccess()
    fun onPinSubmitClearFailure()
    fun onPinSubmitError(throwable: Throwable)
    fun onPinSubmitComplete()
  }

  interface MasterPinCallback {
    fun onMasterPinMissing()
    fun onMasterPinPresent()
  }
}
